content/backend/graphql-scala/6-interfaces.md

```diff
@@ -0,0 +1,352 @@
+
+
+### Vote entity
+
+If you want, you can make similar step for `Vote` entity. And then follow instructions and check everything is ok.
+
+<Instruction>
+
+Create `Vote` class
+
+```scala
+case class Vote(id: Int, userId: Int, linkId: Int, createdAt: DateTime = DateTime.now)
+```
+
+</Instruction>
+
+Database setup.
+
+<Instruction>
+
+Add following content to the `DBSchema` class:
+
+```scala
+class VotesTable(tag: Tag) extends Table[Vote](tag, "VOTES"){
+    def id = column[Int]("ID", O.PrimaryKey)
+    def userId = column[Int]("USER_ID")
+    def linkId = column[Int]("LINK_ID")
+    def createdAt = column[DateTime]("CREATED_AT")
+
+    def * = (id, userId, linkId, createdAt).mapTo[Vote]
+  }
+
+  val Votes = TableQuery[VotesTable]
+```
+
+</Instruction>
+
+If you want to, you can add also foreign keys' definitions but to keep the code simple I decided to not doing it.
+
+<Instruction>
+
+In file `DBSchema` in function databaseSetup: Add an action `Votes.schema.create` at beginning of the sentence and then add few users later in this function:
+
+```scala
+Votes ++= Seq(
+  Vote(1, 1, 1),
+  Vote(2, 1, 2),
+  Vote(3, 1, 3),
+  Vote(4, 2, 2),
+)
+```
+
+</Instruction>
+
+
+Add votes retrieval function.
+
+<Instruction>
+
+In `DAO` class add a function:
+
+```scala
+def getVotes(ids: Seq[Int]): Future[Seq[Vote]] = {
+    db.run(
+      Votes.filter(_.id inSet ids).result
+    )
+}
+```
+
+</Instruction>
+
+GraphQL part:
+
+<Instruction>
+
+In `GraphQLSchema` add :
+
+```scala
+implicit val VoteType = deriveObjectType[Unit, Vote]()
+implicit val voteHasId = HasId[Vote, Int](_.id)
+
+val votesFetcher = Fetcher(
+  (ctx: MyContext, ids: Seq[Int]) => ctx.dao.getVotes(ids)
+)
+```
+
+</Instruction>
+
+Add fetcher to resolvers.
+
+<Instruction>
+
+Add lastly created fetcher to the resolvers list. In the same file, replace constant `Resolver` with:
+
+```
+val Resolver = DeferredResolver.fetchers(linksFetcher, usersFetcher)
+```
+
+</Instruction>
+
+Add fields to main ObjectType:
+
+<Instruction>
+
+Add to `QueryType.fields`:
+
+```scala
+//val Ids = Argument("ids", ListInputType(IntType))
+Field("votes",
+        ListType(VoteType),
+        arguments = List(Ids),
+        resolve = c => votesFetcher.deferSeq(c.arg(Ids))
+)
+```
+
+</Instruction>
+
+Following query now should be able to execute:
+
+```graphql
+
+query {
+
+    votes(ids: [1,2]){
+    	id
+    	createdAt
+  	}
+}  
+```
+
+### Finding a common parts
+
+If you will see once again to the code, you will find out the parts that are very similar. Like `HasId` for all three types:
+
+```scala
+implicit val linkHasId = HasId[Link, Int](_.id)
+implicit val userHasId = HasId[User, Int](_.id)
+implicit val voteHasId = HasId[Vote, Int](_.id)
+```
+
+What if you want to add more entities? You will spread those code even more.
+
+The solution for this is an interface. We can provide an interface that will be extended by any on the entities, and then you can make only one `HasId`, for example.
```

Comment:
`by any on the entities` -> `by any of the entities`
`This way, for example, you will need just one HasId`